File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/player_screen/view/adapter/PlayQueueViewHolder.java
package com.github.anrimian.musicplayer.ui.player_screen.view.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import com.github.anrimian.musicplayer.R;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem;
import com.github.anrimian.musicplayer.ui.common.format.wrappers.CompositionItemWrapper;
import com.github.anrimian.musicplayer.ui.utils.OnPositionItemClickListener;
import com.github.anrimian.musicplayer.ui.utils.OnViewItemClickListener;
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.touch_helper.drag_and_drop.DragListener;

import java.util.List;

import javax.annotation.Nonnull;

/**
 * Created on 31.10.2017.
 */

class PlayQueueViewHolder extends RecyclerView.ViewHolder implements DragListener {

    private final CompositionItemWrapper compositionItemWrapper;

    private PlayQueueItem playQueueItem;

    PlayQueueViewHolder(LayoutInflater inflater,
                        ViewGroup parent,
                        OnPositionItemClickListener<PlayQueueItem> onCompositionClickListener,
                        OnViewItemClickListener<PlayQueueItem> menuClickListener,
                        OnPositionItemClickListener<PlayQueueItem> iconClickListener) {
        super(inflater.inflate(R.layout.item_play_queue, parent, false));
        View btnActionsMenu = itemView.findViewById(R.id.btn_actions_menu);

        compositionItemWrapper = new CompositionItemWrapper(itemView,
                o -> iconClickListener.onItemClick(getBindingAdapterPosition(), playQueueItem),
                composition -> onCompositionClickListener.onItemClick(getBindingAdapterPosition(), playQueueItem)
        );

        btnActionsMenu.setOnClickListener(v -> menuClickListener.onItemClick(v, playQueueItem));
    }

    @Override
    public void onDragStateChanged(boolean dragging) {
        compositionItemWrapper.showAsDraggingItem(dragging);
    }

    void bind(@Nonnull PlayQueueItem item, boolean showCovers) {
        this.playQueueItem = item;
        Composition composition = item.getComposition();
        compositionItemWrapper.bind(composition, showCovers);
//        compositionItemWrapper.showNumber(getAdapterPosition());
    }

    void update(PlayQueueItem item, List<Object> payloads) {
        this.playQueueItem = item;
        compositionItemWrapper.update(item.getComposition(), payloads);
    }

    void release() {
        compositionItemWrapper.release();
    }

    void setCoversVisible(boolean visible) {
        compositionItemWrapper.showCompositionImage(visible);
    }

    void showAsCurrentItem(boolean show) {
        compositionItemWrapper.showAsCurrentComposition(show);
    }

    void showAsPlaying(boolean playing, boolean animate) {
        compositionItemWrapper.showAsPlaying(playing, animate);
    }

    PlayQueueItem getPlayQueueItem() {
        return playQueueItem;
    }


}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/player_screen/queue/adapter/PlayQueueViewHolder.kt
package com.github.anrimian.musicplayer.ui.player_screen.queue.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.anrimian.filesync.models.state.file.FileSyncState
import com.github.anrimian.musicplayer.R
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem
import com.github.anrimian.musicplayer.ui.common.format.wrappers.CompositionItemWrapper
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.mvp.MvpDiffAdapter
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.short_swipe.SwipeListener
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.touch_helper.drag_and_drop.DragListener

/**
 * Created on 31.10.2017.
 */
class PlayQueueViewHolder(
    inflater: LayoutInflater,
    parent: ViewGroup,
    onCompositionClickListener: (Int, PlayQueueItem) -> Unit,
    menuClickListener: (View, PlayQueueItem) -> Unit,
    iconClickListener: (Int, PlayQueueItem) -> Unit
) : MvpDiffAdapter.MvpViewHolder(inflater.inflate(R.layout.item_play_queue, parent, false)),
    DragListener, SwipeListener {

    private val compositionItemWrapper: CompositionItemWrapper<Composition>

    private lateinit var playQueueItem: PlayQueueItem

    init {
        val btnActionsMenu = itemView.findViewById<View>(R.id.btnActionsMenu)
        compositionItemWrapper = CompositionItemWrapper(
            itemView,
            { iconClickListener(bindingAdapterPosition, playQueueItem) }
        ) { onCompositionClickListener(bindingAdapterPosition, playQueueItem) }

        btnActionsMenu.setOnClickListener { v -> menuClickListener(v, playQueueItem) }
    }

    override fun release() {
        compositionItemWrapper.release()
    }

    override fun onDragStateChanged(dragging: Boolean) {
        compositionItemWrapper.showAsDraggingItem(dragging)
    }

    override fun onSwipeStateChanged(swipeOffset: Float) {
        compositionItemWrapper.showAsSwipingItem(swipeOffset)
    }

    fun bind(item: PlayQueueItem, showCovers: Boolean) {
        playQueueItem = item
        compositionItemWrapper.bind(item, showCovers)
    }

    fun update(item: PlayQueueItem, payloads: List<*>) {
        playQueueItem = item
        compositionItemWrapper.update(item, payloads)
    }

    fun setCoversVisible(visible: Boolean) {
        compositionItemWrapper.showCompositionImage(visible)
    }

    fun showAsCurrentItem(show: Boolean) {
        compositionItemWrapper.showAsCurrentComposition(show)
    }

    fun showAsPlaying(playing: Boolean, animate: Boolean) {
        compositionItemWrapper.showAsPlaying(playing, animate)
    }

    fun setFileSyncStates(fileSyncStates: Map<Long, FileSyncState>) {
        compositionItemWrapper.showFileSyncState(fileSyncStates[playQueueItem.id])
    }

    fun getPlayQueueItem() = playQueueItem
}